### PR TITLE
[DependencyInjection] Add support for generating lazy closures

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

On my mission to improve support for closures in DI, this allows generating lazy closures.

What is a lazy closure?

Right now, when we generate a closure, we generate something like:
```php
return ($container->privates['foo'] ?? $container->getFooService())->someMethod(...);
```

As you can see, this requires instantiating the "foo" service, while the closure may never be called.

After this PR, a closure registered as a service can be declared as lazy, which will end up generating this instead (conceptually):
```php
return fn (...$arguments) => ($container->privates['foo'] ?? $container->getFooService())->someMethod(...$arguments);
```

With this code, "foo" is created only if the closure is actually called.

This is opt-in since it requires the "lazy" flag on the service definition.

### CODE REVIEW

**Comment by stof** [comment]
Generating a closure with a variadic argument will not produce the same behavior than the previous code though. `Closure::fromCallable()` or first-class callables are producing a callable that exposes the original signature when using Reflection.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -1187,31 +1214,34 @@ private function doResolveServices(mixed $value, array &$inlineServices = [], bo
             $reference = $value->getValues()[0];
             $value = fn () => $this->resolveServices($reference);
         } elseif ($value instanceof IteratorArgument) {
-            $value = new RewindableGenerator(function () use ($value, &$inlineServices) {
+            $containerRef = \WeakReference::create($this);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: This seems to include a bunch of changes that are unrelated to generating lazy closures AFAICT

**Comment by nicolas-grekas** [comment]
@stof that's right. We could generate a signature from the proxied method when it's known, but that'd work only with PhpDumper, not with ContainerBuilder. But in most cases, the exact signature of a closure doesn't matter, as long at it behaves as expected when being called.

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/lazy_closure.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Gets the public 'closure' shared service.
+     *
+     * @return \Closure
+     */
+    protected static function getClosureService($container, $lazyLoad = true)
+    {
+        $containerRef = $container->ref;
+
+        return $container->services['closure'] = function (...$arguments) use ($containerRef) { static $service; return ($service ??= ($container = $containerRef->get()) ? new \Symfony\Component\DependencyInjection\Tests\Compiler\Foo() : null)->cloneFoo(...$arguments); };
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/lazy_closure.php]: `null->cloneFoo(...$arguments)` will provide a totally WTF error. I think we need to have a much better handling of the case where the container ref is not valid anymore.

**Comment by stof** [comment]
@nicolas-grekas then this requires an explicit documentation of this limitation, to avoid issues for cases where the callable signature matters (for instance if you try to use such callable in a controller resolver, as argument resolvers rely on the signature of the controller callable)

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: moved to #49684

**Comment by nicolas-grekas** [comment]
OK, so you triggered me :) Here is an implementation that respects the target signature (using eval() for ContainerBuilder, as already done for lazy services.)

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/lazy_closure.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/lazy_closure.php]: This won't happen in practice. I wrote it this way to keep a onliner.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 3]
```diff
@@ -1050,13 +1051,48 @@ private function createService(Definition $definition, array &$inlineServices, b
...
+                $callable[0] instanceof Reference
+                || $callable[0] instanceof Definition && !isset($inlineServices[spl_object_hash($callable[0])])
+            )) {
+                $containerRef = $this->containerRef ??= \WeakReference::create($this);
+                $class = ($callable[0] instanceof Reference ? $this->findDefinition($callable[0]) : $callable[0])->getClass();
+                $r = $this->getReflectionClass($class);
+
+                if ($r && $r->hasMethod($callable[1])) {
+                    $signature = ProxyHelper::exportSignature($r->getMethod($callable[1]));
+                    $signature = preg_replace('/[^(]++(\(.*\))(: .*?)?$/', '\1 use ($containerRef, $callable, &$inlineServices)', $signature);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: What happens if the method being proxied has a parameter named `$callable` or `$containerRef` ?

Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: shouldn't the replacement keep `\2` as well ?

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 4]
```diff
@@ -1179,6 +1180,24 @@ private function addNewInstance(Definition $definition, string $return = '', str
...
+                if (['...'] === $arguments && $definition->isLazy() && 'Closure' === ($definition->getClass() ?? 'Closure') && (
+                    $callable[0] instanceof Reference
+                    || ($callable[0] instanceof Definition && !$this->definitionVariables->contains($callable[0]))
+                )) {
+                    $this->addContainerRef = true;
+                    $class = ($callable[0] instanceof Reference ? $this->container->findDefinition($callable[0]) : $callable[0])->getClass();
+                    $r = $this->container->getReflectionClass($class);
+
+                    if ($r && $r->hasMethod($callable[1])) {
+                        $signature = ProxyHelper::exportSignature($r->getMethod($callable[1]));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Does this preserve parameter attributes ?

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: good catch, fixed

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: ok, done

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: It doesn't, and it shouldn't: proxies don't have to repeat the attributes.

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This means that such lazy callables won't be suitable for places doing runtime usages of attributes, like controller argument resolvers.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Correct. That's also the case for other lazy services: generated proxies don't copy attributes. Should be documented.

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Actually, it depends on the kind on proxies being generated. For lazy ghosts, we don't override methods, so they have the attributes in their inherited signatures.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
